pkg/parser/grok/grok.go

```diff
@@ -0,0 +1,164 @@
+func NewConfigWithID(operatorID string) *Config {
+	return &Config{
+		ParserConfig: helper.NewParserConfig(operatorID, operatorType),
+	}
+}
+
+// Config is the configuration of a grok parser operator.
+type Config struct {
+	helper.ParserConfig `mapstructure:",squash"`
+
+	// grok pattern
+	Pattern string `mapstructure:"pattern"`
+
+	Cache struct {
+		Size uint16 `mapstructure:"size"`
+	} `mapstructure:"cache"`
+
+	// array of keys that will be present in the final values
+	Include []string `mapstructure:"include"`
+
+	// array of keys that will be removed from the final values
+	Exclude []string `mapstructure:"exclude"`
+}
+
+// Build will build a grok parser operator.
+func (c Config) Build(logger *zap.SugaredLogger) (operator.Operator, error) {
+	parserOperator, err := c.ParserConfig.Build(logger)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.Pattern == "" {
+		return nil, fmt.Errorf("missing required field 'grok'")
+	}
+
+	g, err := grok.NewWithConfig(&grok.Config{NamedCapturesOnly: true})
+	if err != nil {
+		return nil, fmt.Errorf("new grok object: %w", err)
+	}
+
+	include := map[string]bool{}
+	exclude := map[string]bool{}
+	includeKeysPresent := false
+	excludeKeysPresent := false
+	if c.Include != nil && len(c.Include) > 0 {
+		includeKeysPresent = true
+		for _, k := range c.Include {
+			include[k] = true
+		}
+	}
+	if c.Exclude != nil && len(c.Exclude) > 0 {
+		excludeKeysPresent = true
+		for _, k := range c.Exclude {
+			exclude[k] = true
+		}
+	}
+
+	op := &Parser{
+		ParserOperator:     parserOperator,
+		grok:               g,
+		pattern:            c.Pattern,
+		includeKeysPresent: includeKeysPresent,
+		include:            include,
+		excludeKeysPresent: excludeKeysPresent,
+		exclude:            exclude,
+	}
+
+	if c.Cache.Size > 0 {
+		op.cache = newMemoryCache(c.Cache.Size, 0)
+		logger.Debugf("configured %s with memory cache of size %d", op.ID(), op.cache.maxSize())
+	}
+
+	return op, nil
+}
+
+// Parser is an operator that parses grok in an entry.
+type Parser struct {
+	helper.ParserOperator
+	grok               *grok.Grok
+	pattern            string
+	cache              cache
+	includeKeysPresent bool
+	include            map[string]bool
+	excludeKeysPresent bool
+	exclude            map[string]bool
+}
+
+// Process will parse an entry for grok.
+func (r *Parser) Process(ctx context.Context, entry *entry.Entry) error {
+	return r.ParserOperator.ProcessWith(ctx, entry, r.parse)
+}
+
+// parse will parse a value using the supplied grok.
+func (r *Parser) parse(value interface{}) (interface{}, error) {
+	var raw string
+	switch m := value.(type) {
+	case string:
+		raw = m
+	default:
+		return nil, fmt.Errorf("type '%T' cannot be parsed as grok", value)
+	}
+	return r.match(raw)
+}
+
+func (r *Parser) match(value string) (interface{}, error) {
+	if r.cache != nil {
+		if x := r.cache.get(value); x != nil {
+			return x, nil
+		}
+	}
+
+	values, err := r.grok.Parse(r.pattern, value)
```

Comment:
This looks like all string values, is it considered when the user desires some parsed values to map to appropriate types such as status -> int etc.?

Reply:
Yeah as of now both the regex and this grok parser returns string values. For converting it to different datatype the attribute operator is used 
ex:- https://github.com/SigNoz/nginx-logs-parsing/blob/c14a2178a050293a50549b79991ed3858e295c78/clickhouse-setup/otel-collector-config.yaml#L123

Reply:
I saw some TypedParse or similar in the grok lib API so I was thinking If one would prefer that instead of adding another processor later.

Reply:
Yeah, makes sense. Though if someone uses an existing pattern it might not contain corresponding types. But if they use patterns like number or string then they should be able to do it right here. Let me check and add it.

Reply:
Added the support for this